# Deployement 
Ensure at least 15 seconds of data is collected, trimming edges to obtain 10 seconds of relevant data. Collect 3-5 samples per activity class and report accuracy using both featurized and raw data. You have to train on UCI dataset (You can use the entire dataset if you want) and test it on the data that you have collected and report the accuracy and confusion matrix. Test your model's performance on the collected data, explaining why it succeeded or failed.

In [2]:

from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
import pandas as pd
import numpy as np
import os
import pickle

time = 10
offset = 300
folders = ["LAYING","SITTING","STANDING","WALKING","WALKING_DOWNSTAIRS","WALKING_UPSTAIRS"]
classes = {"WALKING":1,"WALKING_UPSTAIRS":2,"WALKING_DOWNSTAIRS":3,"SITTING":4,"STANDING":5,"LAYING":6}

dataset_dir = os.path.join("Physics_tool_box_data")

X_test=[]
y_test=[]


for folder in folders:
    files = os.listdir(os.path.join(dataset_dir,folder))
    for file in files:

        df = pd.read_csv(os.path.join(dataset_dir,folder,file),sep=",",header=0)
        df = df.iloc[offset:offset+time*100,1:4]
        X_test.append(df.values)
        y_test.append(classes[folder])

#print(X_test)
X_test = np.array(X_test)
y_test = np.array(y_test)

X_test = X_test.reshape(X_test.shape[1]*X_test.shape[0],X_test.shape[2])
y_test = np.repeat(y_test, 1000)



In [3]:
with open('model_raw.pkl', 'rb') as f:
    model_raw = pickle.load(f)
y_predict=model_raw.predict(X_test)


In [4]:
df=pd.DataFrame(np.zeros((6,6),int),columns=['WALKING','WALKING_UPSTAIRS','WALKING_DOWNSTAIRS','SITTING','STANDING','LAYING'],index=['WALKING','WALKING_UPSTAIRS','WALKING_DOWNSTAIRS','SITTING','STANDING','LAYING'])

In [5]:
for i in range(0,len(y_predict)):
    df.iloc[y_predict[i]-1,y_test[i]-1]=df.iloc[y_predict[i]-1,y_test[i]-1]+1
df['PRECISION'] = np.zeros(len(df))
df.loc['RECALL']=np.zeros(7,int)
for i in range(len(df)):
    sum=df.iloc[i,:].sum()
    if sum==0:
        df.iloc[i,6]=0
    else:
        df.iloc[i,6]=df.iloc[i,i]/sum
    
for i in range(len(df.columns)):
    sum=df.iloc[:,i].sum()
    df.iloc[6,i]=df.iloc[i,i]/sum

total_sum=0
sum=0
for i in range(len(df.columns)):
    total_sum=total_sum+df.iloc[:,i].sum()
for i in range(len(df.columns)):
    sum=sum+df.iloc[i,i]
accur=sum/total_sum
print("Accuracy: ",accur)

Accuracy:  0.21086162269799158


In [6]:
df

WALKING  WALKING_UPSTAIRS  WALKING_DOWNSTAIRS  \
WALKING              262.000000           270.000          234.000000   
WALKING_UPSTAIRS     148.000000           174.000          170.000000   
WALKING_DOWNSTAIRS   790.000000           804.000          812.000000   
SITTING              196.000000            94.000           58.000000   
STANDING               0.000000             0.000            0.000000   
LAYING              1604.000000          1658.000         1726.000000   
RECALL                 0.087333             0.058            0.270667   

                    SITTING  STANDING       LAYING  PRECISION  
WALKING                   0         0     0.000000   0.342037  
WALKING_UPSTAIRS         18        12     0.000000   0.333333  
WALKING_DOWNSTAIRS     1090      1034   452.000000   0.162987  
SITTING                   0         0     0.000000   0.000000  
STANDING                  0         0     0.000000   0.000000  
LAYING                 1892      1954  2548.000000   0.223862  
RECALL                    0         0     0.849333   0.000000

In [17]:
# with open('model_features.pkl', 'rb') as f:
#     model_features = pickle.load(f)
# y_predict=model_raw.predict(X_test)
import tsfel 
from sklearn.preprocessing import StandardScaler
y_list = []

features_list = []

data_directory = os.path.join("Physics_tool_box_data")
# Iterate through each activity folder
for activity_folder in os.listdir(data_directory):
    activity_path = os.path.join(data_directory, activity_folder)

    # Iterate through files in the activity folder
    for filename in os.listdir(activity_path):
        file_path = os.path.join(activity_path, filename)

        # Read the data from the file
        data = pd.read_csv(file_path)
        data=data.iloc[:,1:4]
        #y_list.append(activity_folder)
        # Extract features using TSFEL
        cfg = tsfel.get_features_by_domain('statistical')
        # extracted_features = tsfel.time_series_features_extractor(cfg, data)
        
        # df = pd.DataFrame(X)
        fs = 50
        
        # displaying the DataFrame
        extracted_features = tsfel.time_series_features_extractor(cfg, data,fs=fs,window_size=100)
        
        # Append the features DataFrame to the list
        features_list.append(extracted_features)
        y_list.extend([classes[activity_folder]]*len(extracted_features))
        # Concatenate all features DataFrames into one
        features_df = pd.concat(features_list, ignore_index=True, axis=0)

scaler = StandardScaler()
scaled_data = scaler.fit_transform(features_df)
scaled_data

*** Feature extraction started ***



*** Feature extraction finished ***
*** Feature extraction started ***



*** Feature extraction finished ***
*** Feature extraction started ***



*** Feature extraction finished ***
*** Feature extraction started ***



*** Feature extraction finished ***
*** Feature extraction started ***



*** Feature extraction finished ***
*** Feature extraction started ***



*** Feature extraction finished ***
*** Feature extraction started ***



*** Feature extraction finished ***
*** Feature extraction started ***



*** Feature extraction finished ***
*** Feature extraction started ***



*** Feature extraction finished ***
*** Feature extraction started ***



*** Feature extraction finished ***
*** Feature extraction started ***



*** Feature extraction finished ***
*** Feature extraction started ***



*** Feature extraction finished ***
*** Feature extraction started ***



*** Feature extraction finished ***
*** Feature extraction started ***



*** Feature extraction finished ***
*** Feature extraction started ***



*** Feature extraction finished ***
*** Feature extraction started ***



*** Feature extraction finished ***
*** Feature extraction started ***



*** Feature extraction finished ***
*** Feature extraction started ***



*** Feature extraction finished ***


array([[-0.51138092, -0.51138092,  0.        , ..., -0.93236962,
        -0.82073136, -0.69360613],
       [-0.51223453, -0.51223453,  0.        , ..., -3.77692427,
        -0.86403408, -0.69860818],
       [-0.51163344, -0.51163344,  0.        , ...,  0.67409795,
        -0.92148814, -0.70321676],
       ...,
       [-0.2019435 , -0.2019435 ,  0.        , ...,  0.60771445,
         0.71072179,  0.32822904],
       [ 0.06612068,  0.06612068,  0.        , ..., -0.49665654,
         0.93065361,  0.60992509],
       [-0.29478442, -0.29478442,  0.        , ...,  1.0224739 ,
         0.18455707, -0.20816283]])

In [18]:
print("scale data: ",scaled_data.shape)

scale data:  (293, 120)


In [19]:
with open('model_features.pkl', 'rb') as f:
    model_features = pickle.load(f)
y_predict=model_features.predict(scaled_data)
y_predict


array([4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4,
       4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4,
       4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4,
       4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4,
       4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4,
       4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4,
       4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4,
       4, 4, 3, 3, 3, 3, 3, 2, 2, 3, 3, 3, 3, 3, 2, 2, 4, 4, 4, 2, 4, 2,
       1, 1, 1, 4, 3, 3, 1, 1, 3, 1, 1, 4, 3, 2, 1, 2, 2, 3, 1, 3, 3, 1,
       1, 1, 3, 1, 2, 2, 2, 3, 3, 1, 3, 2, 2, 4, 2, 1, 2, 2, 2, 4, 2, 2,
       4, 2, 2, 4, 4, 3, 2, 2, 2, 2, 2, 4, 1, 2, 4, 2, 2, 4, 3, 2, 2, 2,
       4, 4, 2, 2, 3, 2, 1, 2, 2, 1, 2, 3, 2, 2, 4, 2, 2, 2, 4, 4, 5, 2,
       2, 4, 4, 2, 2, 2, 1, 2, 2, 4, 2, 2, 2, 4, 4, 2, 2, 2, 2, 3, 1, 1,
       2, 1, 2, 4, 4, 2, 1])

In [20]:
df=pd.DataFrame(np.zeros((6,6),int),columns=['WALKING','WALKING_UPSTAIRS','WALKING_DOWNSTAIRS','SITTING','STANDING','LAYING'],index=['WALKING','WALKING_UPSTAIRS','WALKING_DOWNSTAIRS','SITTING','STANDING','LAYING'])

In [21]:
for i in range(0,len(y_predict)):
    df.iloc[y_predict[i]-1,y_list[i]-1]=df.iloc[y_predict[i]-1,y_list[i]-1]+1
df['PRECISION'] = np.zeros(len(df))
df.loc['RECALL']=np.zeros(7,int)
for i in range(len(df)):
    sum=df.iloc[i,:].sum()
    if sum==0:
        df.iloc[i,6]=0
    else:
        df.iloc[i,6]=df.iloc[i,i]/sum
    
for i in range(len(df.columns)):
    sum=df.iloc[:,i].sum()
    df.iloc[6,i]=df.iloc[i,i]/sum

total_sum=0
sum=0
for i in range(len(df.columns)):
    total_sum=total_sum+df.iloc[:,i].sum()
for i in range(len(df.columns)):
    sum=sum+df.iloc[i,i]
accur=sum/total_sum
print("Accuracy: ",accur)

Accuracy:  0.3103391770963009


In [22]:
df

WALKING  WALKING_UPSTAIRS  WALKING_DOWNSTAIRS  SITTING  \
WALKING             13.000000          7.000000            3.000000        0   
WALKING_UPSTAIRS     9.000000         26.000000           26.000000        0   
WALKING_DOWNSTAIRS  18.000000          2.000000            6.000000        0   
SITTING              8.000000         10.000000           10.000000       47   
STANDING             0.000000          1.000000            0.000000        0   
LAYING               0.000000          0.000000            0.000000        0   
RECALL               0.270833          0.565217            0.133333        1   

                    STANDING  LAYING  PRECISION  
WALKING                    0       0   0.565217  
WALKING_UPSTAIRS           0       0   0.426230  
WALKING_DOWNSTAIRS         0       0   0.230769  
SITTING                   55      52   0.258242  
STANDING                   0       0   0.000000  
LAYING                     0       0   0.000000  
RECALL                     0       0   0.000000

We are getting approx 21% and 31% accuracy for model trained on raw data and featurized data respectively. Although model gives 63% and 89% accuracy for UCI dataset test data. This difference might occurred due to error while taking reading, Although we took reading carefully by following all instructions. 